scripts/test.py

```diff
@@ -218,138 +192,214 @@ def run_test(driver: Path, log_queue: queue.Queue) -> int:
+    """
+    # None means that stdout and stderr are handled by parent, i.e., they go to console by default
+    stdout = None
+    stderr = None
+
+    if silent:
+        stdout = subprocess.DEVNULL
+        stderr = subprocess.DEVNULL
+    elif log_path:
+        stdout = open(f"{log_path}.stdout.log", "w")
+        stderr = open(f"{log_path}.stderr.log", "w")
+
+    try:
+        subprocess.run(cmd, env=env, stdout=stdout, stderr=stderr, timeout=timeout, check=True)
+    except subprocess.CalledProcessError as e:
+        return e.returncode, f"{e.cmd} failed with return code {e.returncode}", False
+    except subprocess.TimeoutExpired as e:
+        return TIMEOUT_RETURNCODE, f"{e.cmd} took more than {e.timeout}", True
+    return 0, "", False
+
+def clean() -> bool:
+    print("Cleaning project...")
+    return_code, error_msg, _ = run_subprocess(
+        cmd=["make", "-C", PROJECT_LOCATION, "clean"],
+        timeout=BUILD_TIMEOUT_SECONDS,
+        silent=True,
     )
+    print("")
     if return_code != 0:
-        return return_code
+        print("Error when cleaning:", error_msg)
+        return False
+    return True
+
+def make(with_coverage, silent) -> bool:
+    print("Running make...\n")
+
+    cmd = ["make", "-C", PROJECT_LOCATION, "bin/c_compiler"]
+    if with_coverage:
+        # Run coverage if needed
+        print("Making with coverage...")
+        shutil.rmtree(COVERAGE_FOLDER, ignore_errors=True)
+        cmd = ["make", "-C", PROJECT_LOCATION, "with_coverage"]
 
-    init_print_message, init_xml_message = init_message
-    log_queue.put((
-        f"{init_print_message}\t> " + Fore.GREEN + "Pass",
-        f"{init_xml_message}</testcase>\n"
-    ))
+    return_code, error_msg, _ = run_subprocess(cmd=cmd, timeout=BUILD_TIMEOUT_SECONDS, silent=silent)
 
-    return 0
+    print("")
+    if return_code != 0:
+        print("Error when making:", error_msg)
+        return False
 
+    return True
 
-def empty_log_queue(
-    log_queue: queue.Queue,
+def process_result(
+    result: Result,
+    xml_file: JUnitXMLFile,
     verbose: bool = False,
-    progress_bar: ProgressBar = None
+    progress_bar: ProgressBar = None,
 ):
-    """
-    Empty log queue while logs are sent to XML file and terminal/progress bar
-    """
-    while not log_queue.empty():
-        print_msg, xml_message = log_queue.get()
-        processed_msg = re.sub(r"(/workspaces/).+?/([compiler_tests|bin])", r"\1\2", print_msg)
+    xml_file.write(result.to_xml())
 
-        if verbose:
-            print(print_msg)
-        else:
-            if "Pass" in processed_msg:
-                progress_bar.test_passed()
-            elif "Fail" in processed_msg:
-                progress_bar.test_failed()
+    if verbose:
+        print(result.to_log())
+        return
+
+    if not progress_bar:
+        return
+
+    if result.passed:
+        progress_bar.test_passed()
+    else:
+        progress_bar.test_failed()
```

Comment:
I think that's cleaner than early return as they are logically linked
```suggestion
    if progress_bar:
      if result.passed:
          progress_bar.test_passed()
      else:
          progress_bar.test_failed()
```